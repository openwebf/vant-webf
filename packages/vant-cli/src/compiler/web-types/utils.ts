// myName -> my-name
export function toKebabCase(input: string): string {
  return input.replace(
    /[A-Z]/g,
    (val, index) => (index === 0 ? '' : '-') + val.toLowerCase()
  );
}

// name `v2.0.0` -> name
export function removeVersion(str: string) {
  return str.replace(/`(\w|\.)+`/g, '').trim();
}

// *boolean* -> boolean
// _boolean_ -> boolean
export function formatType(type: string) {
  return type.replace(/(^(\*|_))|((\*|_)$)/g, '');
}

export function normalizePath(path: string): string {
  return path.replace(/\\/g, '/');
}

// `default` `primary` -> ['default', 'primary']
export function formatOptions(options?: string) {
  if (!options) return [];
  return options.replace(/`/g, '').split(' ');
}
