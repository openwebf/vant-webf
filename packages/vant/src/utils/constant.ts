import type { InjectionKey } from 'vue';
import type { FormProvide } from '../form/types';

export const BORDER = 'van-hairline';
export const BORDER_TOP = `${BORDER}--top`;
export const BORDER_LEFT = `${BORDER}--left`;
export const BORDER_BOTTOM = `${BORDER}--bottom`;
export const BORDER_SURROUND = `${BORDER}--surround`;
export const BORDER_TOP_BOTTOM = `${BORDER}--top-bottom`;
export const BORDER_UNSET_TOP_BOTTOM = `${BORDER}-unset--top-bottom`;

export const HAPTICS_FEEDBACK = 'van-haptics-feedback';

export const FORM_KEY: InjectionKey<FormProvide> = Symbol('van-form');
