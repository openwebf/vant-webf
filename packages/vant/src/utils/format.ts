import type { CSSProperties } from 'vue';
import { inBrowser, type Numeric } from './basic';
import { windowWidth, windowHeight } from './dom';
import { isDef, isNumeric } from './validate';

export function addUnit(value?: Numeric): string | undefined {
  if (isDef(value)) {
    return isNumeric(value) ? `${value}px` : String(value);
  }
  return undefined;
}

export function getSizeStyle(
  originSize?: Numeric | Numeric[]
): CSSProperties | undefined {
  if (isDef(originSize)) {
    if (Array.isArray(originSize)) {
      return {
        width: addUnit(originSize[0]),
        height: addUnit(originSize[1]),
      };
    }
    const size = addUnit(originSize);
    return {
      width: size,
      height: size,
    };
  }
}

export function getZIndexStyle(zIndex?: Numeric) {
  const style: CSSProperties = {};
  if (zIndex !== undefined) {
    style.zIndex = +zIndex;
  }
  return style;
}

// cache
let rootFontSize: number;

function getRootFontSize() {
  if (!rootFontSize) {
    const doc = document.documentElement;
    const fontSize =
      doc.style.fontSize || window.getComputedStyle(doc).fontSize;

    rootFontSize = parseFloat(fontSize);
  }

  return rootFontSize;
}

function convertRem(value: string) {
  value = value.replace(/rem/g, '');
  return +value * getRootFontSize();
}

function convertVw(value: string) {
  value = value.replace(/vw/g, '');
  return (+value * windowWidth.value) / 100;
}

function convertVh(value: string) {
  value = value.replace(/vh/g, '');
  return (+value * windowHeight.value) / 100;
}

export function unitToPx(value: Numeric): number {
  if (typeof value === 'number') {
    return value;
  }

  if (inBrowser) {
    if (value.includes('rem')) {
      return convertRem(value);
    }
    if (value.includes('vw')) {
      return convertVw(value);
    }
    if (value.includes('vh')) {
      return convertVh(value);
    }
  }

  return parseFloat(value);
}

const camelizeRE = /-(\w)/g;

export const camelize = (str: string): string =>
  str.replace(camelizeRE, (_, c) => c.toUpperCase());

export const kebabCase = (str: string) =>
  str
    .replace(/([A-Z])/g, '-$1')
    .toLowerCase()
    .replace(/^-/, '');

export function padZero(num: Numeric, targetLength = 2): string {
  let str = num + '';

  while (str.length < targetLength) {
    str = '0' + str;
  }

  return str;
}

/** clamps number within the inclusive lower and upper bounds */
export const clamp = (num: number, min: number, max: number): number =>
  Math.min(Math.max(num, min), max);

function trimExtraChar(value: string, char: string, regExp: RegExp) {
  const index = value.indexOf(char);

  if (index === -1) {
    return value;
  }

  if (char === '-' && index !== 0) {
    return value.slice(0, index);
  }

  return value.slice(0, index + 1) + value.slice(index).replace(regExp, '');
}

export function formatNumber(
  value: string,
  allowDot = true,
  allowMinus = true
) {
  if (allowDot) {
    value = trimExtraChar(value, '.', /\./g);
  } else {
    value = value.split('.')[0];
  }

  if (allowMinus) {
    value = trimExtraChar(value, '-', /-/g);
  } else {
    value = value.replace(/-/, '');
  }

  const regExp = allowDot ? /[^-0-9.]/g : /[^-0-9]/g;

  return value.replace(regExp, '');
}

// add num and avoid float number
export function addNumber(num1: number, num2: number) {
  const cardinal = 10 ** 10;
  return Math.round((num1 + num2) * cardinal) / cardinal;
}
