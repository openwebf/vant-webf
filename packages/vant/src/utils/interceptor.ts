import { noop } from './basic';
import { isPromise } from './validate';

export type Interceptor = (
  ...args: any[]
) => Promise<boolean> | boolean | undefined | void;

export function callInterceptor(
  interceptor: Interceptor | undefined,
  {
    args = [],
    done,
    canceled,
  }: {
    args?: unknown[];
    done: () => void;
    canceled?: () => void;
  }
) {
  if (interceptor) {
    // eslint-disable-next-line prefer-spread
    const returnVal = interceptor.apply(null, args);

    if (isPromise(returnVal)) {
      returnVal
        .then((value) => {
          if (value) {
            done();
          } else if (canceled) {
            canceled();
          }
        })
        .catch(noop);
    } else if (returnVal) {
      done();
    } else if (canceled) {
      canceled();
    }
  } else {
    done();
  }
}
