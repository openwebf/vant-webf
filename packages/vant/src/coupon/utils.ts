import { padZero } from '../utils';

export function getDate(timeStamp: number) {
  const date = new Date(timeStamp * 1000);
  return `${date.getFullYear()}.${padZero(date.getMonth() + 1)}.${padZero(
    date.getDate()
  )}`;
}

export const formatDiscount = (discount: number) =>
  (discount / 10).toFixed(discount % 10 === 0 ? 0 : 1);

export const formatAmount = (amount: number) =>
  (amount / 100).toFixed(amount % 100 === 0 ? 0 : amount % 10 === 0 ? 1 : 2);
