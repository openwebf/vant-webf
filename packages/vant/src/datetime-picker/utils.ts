import { extend } from '../utils';
import { pickerSharedProps } from '../picker/Picker';
import type { PropType } from 'vue';
import type { PickerInstance } from '../picker';
import type { DatetimePickerColumnType } from './types';

export const sharedProps = extend({}, pickerSharedProps, {
  filter: Function as PropType<(type: string, values: string[]) => string[]>,
  columnsOrder: Array as PropType<DatetimePickerColumnType[]>,
  formatter: {
    type: Function as PropType<(type: string, value: string) => string>,
    default: (type: string, value: string) => value,
  },
});

export const pickerInheritKeys = Object.keys(pickerSharedProps) as Array<
  keyof typeof pickerSharedProps
>;

export function times<T>(n: number, iteratee: (index: number) => T) {
  if (n < 0) {
    return [];
  }

  const result: T[] = Array(n);

  let index = -1;
  while (++index < n) {
    result[index] = iteratee(index);
  }

  return result;
}

export function getTrueValue(value: string | undefined): number {
  if (!value) {
    return 0;
  }

  while (Number.isNaN(parseInt(value, 10))) {
    if (value.length > 1) {
      value = value.slice(1);
    } else {
      return 0;
    }
  }

  return parseInt(value, 10);
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

// https://github.com/vant-ui/vant/issues/10013
export const proxyPickerMethods = (
  picker: PickerInstance,
  callback: () => void
) => {
  const methods = [
    'setValues',
    'setIndexes',
    'setColumnIndex',
    'setColumnValue',
  ];
  return new Proxy(picker, {
    get: (target, prop: keyof PickerInstance) => {
      if (methods.includes(prop)) {
        return (...args: unknown[]) => {
          target[prop](...args);
          callback();
        };
      }
      return target[prop];
    },
  });
};
