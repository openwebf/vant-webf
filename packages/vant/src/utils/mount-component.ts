import { createApp, reactive, Component } from 'vue';
import { extend } from '../utils';
import { useExpose } from '../composables/use-expose';

export function usePopupState() {
  const state = reactive<{
    show: boolean;
    [key: string]: any;
  }>({
    show: false,
  });

  const toggle = (show: boolean) => {
    state.show = show;
  };

  const open = (props: Record<string, any>) => {
    extend(state, props, { transitionAppear: true });
    toggle(true);
  };

  const close = () => toggle(false);

  useExpose({ open, close, toggle });

  return {
    open,
    close,
    state,
    toggle,
  };
}

export function mountComponent(RootComponent: Component) {
  const app = createApp(RootComponent);
  const root = document.createElement('div');

  document.body.appendChild(root);

  return {
    instance: app.mount(root),
    unmount() {
      app.unmount();
      document.body.removeChild(root);
    },
  };
}
