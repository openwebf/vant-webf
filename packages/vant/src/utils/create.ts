import { get } from './basic';
import { camelize } from './format';
import { isFunction } from './validate';
import locale from '../locale';

export function createTranslate(name: string) {
  const prefix = camelize(name) + '.';

  return (path: string, ...args: unknown[]) => {
    const messages = locale.messages();
    const message = get(messages, prefix + path) || get(messages, path);

    return isFunction(message) ? message(...args) : message;
  };
}

export type Translate = ReturnType<typeof createTranslate>;

export type Mod = string | { [key: string]: any };
export type Mods = Mod | Mod[];

function genBem(name: string, mods?: Mods): string {
  if (!mods) {
    return '';
  }

  if (typeof mods === 'string') {
    return ` ${name}--${mods}`;
  }

  if (Array.isArray(mods)) {
    return (mods as Mod[]).reduce<string>(
      (ret, item) => ret + genBem(name, item),
      ''
    );
  }

  return Object.keys(mods).reduce(
    (ret, key) => ret + (mods[key] ? genBem(name, key) : ''),
    ''
  );
}

/**
 * bem helper
 * b() // 'button'
 * b('text') // 'button__text'
 * b({ disabled }) // 'button button--disabled'
 * b('text', { disabled }) // 'button__text button__text--disabled'
 * b(['disabled', 'primary']) // 'button button--disabled button--primary'
 */
export function createBEM(name: string) {
  return (el?: Mods, mods?: Mods): Mods => {
    if (el && typeof el !== 'string') {
      mods = el;
      el = '';
    }

    el = el ? `${name}__${el}` : name;

    return `${el}${genBem(el, mods)}`;
  };
}

export type BEM = ReturnType<typeof createBEM>;

export function createNamespace(name: string) {
  const prefixedName = `van-${name}`;
  return [
    prefixedName,
    createBEM(prefixedName),
    createTranslate(prefixedName),
  ] as const;
}
