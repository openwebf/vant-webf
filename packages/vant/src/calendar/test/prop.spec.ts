import { Calendar } from '..';
import { mount, later } from '../../../test';
import { minDate, maxDate } from './utils';

test('should limit max range when using max-range prop and type is range', async () => {
  const wrapper = mount(Calendar, {
    props: {
      type: 'range',
      minDate,
      maxDate,
      maxRange: 3,
      poppable: false,
      lazyRender: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days[12].trigger('click');
  days[18].trigger('click');

  expect(wrapper.emitted<[Date]>('select')![0][0]).toEqual([
    new Date(2010, 0, 13),
  ]);
  expect(wrapper.emitted<[Date]>('select')![1][0]).toEqual([
    new Date(2010, 0, 13),
    new Date(2010, 0, 15),
  ]);
  expect(wrapper.emitted('confirm')).toBeFalsy();
});

test('should limit max range when using max-range prop and type is multiple', async () => {
  const wrapper = mount(Calendar, {
    props: {
      type: 'multiple',
      minDate,
      maxDate,
      maxRange: 2,
      poppable: false,
      showConfirm: false,
      lazyRender: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days[13].trigger('click');
  days[14].trigger('click');
  days[15].trigger('click');

  expect(wrapper.emitted<[Date]>('select')![0][0]).toHaveLength(2);
});

test('show-title prop', async () => {
  const wrapper = mount(Calendar, {
    props: {
      show: true,
    },
  });

  expect(wrapper.find('.van-calendar__header-title').exists()).toBeTruthy();
  await wrapper.setProps({ showTitle: false });
  expect(wrapper.find('.van-calendar__header-title').exists()).toBeFalsy();
});

test('show-subtitle prop', async () => {
  const wrapper = mount(Calendar, {
    props: {
      show: true,
    },
  });

  expect(wrapper.find('.van-calendar__header-subtitle').exists()).toBeTruthy();
  await wrapper.setProps({ showSubtitle: false });
  expect(wrapper.find('.van-calendar__header-subtitle').exists()).toBeFalsy();
});

test('hide close icon when there is no title', async () => {
  const wrapper = mount(Calendar, {
    props: {
      show: true,
    },
  });

  expect(wrapper.find('.van-popup__close-icon').exists()).toBeTruthy();

  await wrapper.setProps({
    showTitle: false,
    showSubtitle: false,
  });
  expect(wrapper.find('.van-popup__close-icon').exists()).toBeFalsy();
});

test('allow-same-day prop', async () => {
  const onSelect = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      type: 'range',
      minDate,
      maxDate,
      poppable: false,
      lazyRender: false,
      onSelect,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days[9].trigger('click');
  days[9].trigger('click');

  expect(onSelect).toHaveBeenLastCalledWith([minDate]);

  await wrapper.setProps({
    allowSameDay: true,
  });

  days[9].trigger('click');
  expect(onSelect).toHaveBeenLastCalledWith([minDate, minDate]);
});

test('min-date after current time', () => {
  const wrapper = mount(Calendar, {
    props: {
      poppable: false,
      minDate: new Date(2200, 0, 1),
      maxDate: new Date(2200, 0, 2),
      lazyRender: false,
    },
  });

  wrapper.find('.van-calendar__confirm').trigger('click');
  expect(wrapper.emitted<[Date]>('confirm')![0][0]).toEqual(
    new Date(2200, 0, 1)
  );
});

test('min-date before current time', () => {
  const wrapper = mount(Calendar, {
    props: {
      poppable: false,
      minDate: new Date(1800, 0, 1),
      maxDate: new Date(1800, 0, 2),
      lazyRender: false,
    },
  });

  wrapper.find('.van-calendar__confirm').trigger('click');
  expect(wrapper.emitted<[Date]>('confirm')![0][0]).toEqual(
    new Date(1800, 0, 2)
  );
});

test('lazy-render prop', () => {
  const wrapper = mount(Calendar, {
    props: {
      minDate,
      maxDate,
      poppable: false,
      lazyRender: false,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('month-show event', async () => {
  const wrapper = mount(Calendar, {
    props: {
      show: true,
    },
  });
  await later(200);

  expect(wrapper.emitted('month-show')).toBeTruthy();
});

test('first day of week', async () => {
  const wrapper = mount(Calendar, {
    props: {
      poppable: false,
      defaultDate: new Date(2020, 7, 1),
      minDate: new Date(2020, 7, 1),
      maxDate: new Date(2020, 7, 30),
      firstDayOfWeek: 2,
      lazyRender: false,
    },
  });

  await later();

  expect(wrapper.findAll('.van-calendar__weekday')[0].text()).toEqual('Tue');

  const day = wrapper.find(
    '.van-calendar__month:first-of-type .van-calendar__day'
  );
  expect(day.text()).toEqual('1');
  expect(day.attributes('style')).toContain(`margin-left: ${(100 * 4) / 7}%`);
});

test('readonly prop', async () => {
  const wrapper = mount(Calendar, {
    props: {
      type: 'multiple',
      minDate,
      maxDate,
      readonly: true,
      poppable: false,
      lazyRender: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days[13].trigger('click');
  expect(wrapper.emitted('select')).toBeFalsy();

  await wrapper.setProps({ readonly: false });
  days[13].trigger('click');
  expect(wrapper.emitted('select')).toBeTruthy();
});

test('should disabled prompt when using show-range-prompt prop', async () => {
  document.getElementsByTagName('html')[0].innerHTML = '';
  const wrapper = mount(Calendar, {
    props: {
      type: 'range',
      minDate,
      maxDate,
      maxRange: 3,
      poppable: false,
      lazyRender: false,
      showRangePrompt: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days[12].trigger('click');
  days[18].trigger('click');

  await later();
  expect(document.querySelector('.van-toast')).toBeFalsy();
});

test('should emit over-range when exceeded max range', async () => {
  const onOverRange = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      type: 'range',
      minDate,
      maxDate,
      maxRange: 3,
      onOverRange,
      poppable: false,
      lazyRender: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days[12].trigger('click');
  days[18].trigger('click');

  expect(onOverRange).toHaveBeenCalledTimes(1);
});

test('should allow default date to be minDate when using allowSameDay prop', () => {
  const minDate = new Date(1800, 0, 1);
  const maxDate = new Date(1800, 0, 29);
  const wrapper = mount(Calendar, {
    props: {
      type: 'range',
      poppable: false,
      minDate,
      maxDate,
      defaultDate: [minDate, minDate],
      lazyRender: false,
      allowSameDay: true,
    },
  });

  wrapper.find('.van-calendar__confirm').trigger('click');
  expect(wrapper.emitted<[Date]>('confirm')![0][0]).toEqual([minDate, minDate]);
});

test('should allow default date to be maxDate when using allowSameDay prop', () => {
  const minDate = new Date(1800, 0, 1);
  const maxDate = new Date(1800, 0, 29);
  const wrapper = mount(Calendar, {
    props: {
      type: 'range',
      poppable: false,
      minDate,
      maxDate,
      defaultDate: [maxDate, maxDate],
      lazyRender: false,
      allowSameDay: true,
    },
  });

  wrapper.find('.van-calendar__confirm').trigger('click');
  expect(wrapper.emitted<[Date]>('confirm')![0][0]).toEqual([maxDate, maxDate]);
});
