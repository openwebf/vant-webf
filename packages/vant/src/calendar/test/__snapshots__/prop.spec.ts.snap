// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazy-render prop 1`] = `
<div class="van-calendar">
  <div class="van-calendar__header">
    <div class="van-calendar__header-title">
      Calendar
    </div>
    <div class="van-calendar__header-subtitle">
    </div>
    <div class="van-calendar__weekdays">
      <span class="van-calendar__weekday">
        Sun
      </span>
      <span class="van-calendar__weekday">
        Mon
      </span>
      <span class="van-calendar__weekday">
        Tue
      </span>
      <span class="van-calendar__weekday">
        Wed
      </span>
      <span class="van-calendar__weekday">
        Thu
      </span>
      <span class="van-calendar__weekday">
        Fri
      </span>
      <span class="van-calendar__weekday">
        Sat
      </span>
    </div>
  </div>
  <div class="van-calendar__body">
    <div class="van-calendar__month">
      <div role="grid"
           class="van-calendar__days"
      >
        <div class="van-calendar__month-mark">
          1
        </div>
        <div role="gridcell"
             style="margin-left: 71.42857142857143%;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          1
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          2
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          3
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          4
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          5
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          6
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          7
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          8
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          9
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          10
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          11
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          12
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          13
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          14
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          15
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          16
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          17
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          18
        </div>
        <div role="gridcell"
             class="van-calendar__day"
             tabindex="-1"
        >
          19
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--selected"
             tabindex="-1"
        >
          <div class="van-calendar__selected-day">
            20
          </div>
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          21
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          22
        </div>
        <div role="gridcell"
             class="van-calendar__day van-calendar__day--disabled"
        >
          23
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          24
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          25
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          26
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          27
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          28
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          29
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          30
        </div>
        <div role="gridcell"
             style="margin-bottom: 0px;"
             class="van-calendar__day van-calendar__day--disabled"
        >
          31
        </div>
      </div>
    </div>
  </div>
  <div class="van-calendar__footer van-safe-area-bottom">
    <button type="button"
            class="van-button van-button--danger van-button--normal van-button--block van-button--round van-calendar__confirm"
    >
      <div class="van-button__content">
        <span class="van-button__text">
          Confirm
        </span>
      </div>
    </button>
  </div>
</div>
`;
